### PR TITLE
fix objc_msgSend.x86-32.S for win32

### PR DESCRIPTION
**Pull request by DHowett-MSFT**

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
This does not include the change for `NSConcreteGlobalBlock`, as I'd prefer to find a better solution.

File: objc_msgSend.x86-32.S [hunk 1]
```diff
@@ -65,8 +65,14 @@
 7:
 	popl  %ebx;
 8:
+#if __ELF__
+	# ELF can support GOT-relative addressing;
+	# PE/COFF and Mach-O need a text relocation.
```
Comment (davidchisnall) [inline objc_msgSend.x86-32.S]: Are you sure about Mach-O?  The `objc_msgSend` tests all pass on macOS.

File: objc_msgSend.x86-32.S [hunk 2]
```diff
@@ -54,7 +54,7 @@
 	push  %ecx                           # _cmd
 	push  %eax                           # &self
 	.cfi_def_cfa_offset 12
-	call  slowMsgLookup@PLT
+	call  CDECL(slowMsgLookup)@PLT
```
Comment (davidchisnall) [inline objc_msgSend.x86-32.S]: Is this definitely meant to be `@PLT` for PE/COFF?

**Comment by DHowett-MSFT** [comment]
Both good questions.

## PE

I think clang strips or doesn't emit something for the errant `@PLT`; it's not necessary but it's also not strictly correct.

## mach-o

I looked at the difference in generated code before/after this fix on OS X (x86-32), and it looks like:

### relocs, before
```
objc_msgSend.o: file format Mach-O 32-bit i386

RELOCATION RECORDS FOR [__text]:
0000015a GENERIC_RELOC_VANILLA SmallObjectClasses
00000154 GENERIC_RELOC_VANILLA _GLOBAL_OFFSET_TABLE_
00000142 GENERIC_RELOC_VANILLA slowMsgLookup
000000e3 GENERIC_RELOC_VANILLA SmallObjectClasses
000000dd GENERIC_RELOC_VANILLA _GLOBAL_OFFSET_TABLE_
000000cb GENERIC_RELOC_VANILLA slowMsgLookup
0000006c GENERIC_RELOC_VANILLA SmallObjectClasses
00000066 GENERIC_RELOC_VANILLA _GLOBAL_OFFSET_TABLE_
00000054 GENERIC_RELOC_VANILLA slowMsgLookup

RELOCATION RECORDS FOR [__compact_unwind]:
00000028 GENERIC_RELOC_VANILLA __text
00000014 GENERIC_RELOC_VANILLA __text
00000000 GENERIC_RELOC_VANILLA __text
```

### relocs, after
```
RELOCATION RECORDS FOR [__text]:
00000148 GENERIC_RELOC_VANILLA _SmallObjectClasses
00000136 GENERIC_RELOC_VANILLA _slowMsgLookup
000000d7 GENERIC_RELOC_VANILLA _SmallObjectClasses
000000c5 GENERIC_RELOC_VANILLA _slowMsgLookup
00000066 GENERIC_RELOC_VANILLA _SmallObjectClasses
00000054 GENERIC_RELOC_VANILLA _slowMsgLookup

RELOCATION RECORDS FOR [__compact_unwind]:
00000028 GENERIC_RELOC_VANILLA __text
00000014 GENERIC_RELOC_VANILLA __text
00000000 GENERIC_RELOC_VANILLA __text
```

### generated code
```diff
--- ms1.asm     2018-03-03 11:02:21.000000000 -0800
+++ ms2.asm     2018-03-03 11:02:10.000000000 -0800
@@ -30,14 +30,13 @@
 8d 44 24 04    leal    4(%esp), %eax
 51     pushl   %ecx
 50     pushl   %eax
 e8 31 ff ff ff         calll   -207 <_objc_msgSend_fpret>
 83 c4 08       addl    $8, %esp
 ff e0  jmpl    *%eax
 53     pushl   %ebx
 e8 00 00 00 00         calll   0 <_objc_msgSend+0x65>
 5b     popl    %ebx
-81 c3 03 00 00 00      addl    $3, %ebx
-8d 83 00 00 00 00      leal    (%ebx), %eax
+8d 05 00 00 00 00      leal    0, %eax
 8b 00  movl    (%eax), %eax
 5b     popl    %ebx
 eb 9a  jmp     -102 <_objc_msgSend+0x11>
```

(excerpted from objc_msgSend only)

**Comment by DHowett-MSFT** [comment]
It looks like either way, mach-o requires a textrel; this just changes what that relocation is to. The tests seem to pass regardless.

**Pull request merged**